runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -187,11 +217,152 @@ public void flatMap(T t, Collector<T> collector) throws Exception {
+                      windowingStrategyProto,
+                      rehydratedComponents);
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalStateException(
+              String.format("Unable to hydrate GroupByKey windowing strategy %s.",
+                      windowingStrategyProto),
+              e);
+    }
+
+
+    SingletonKeyedWorkItemCoder<K, V> workItemCoder = SingletonKeyedWorkItemCoder.of(
+            inputCoder.getKeyCoder(),
+            inputCoder.getValueCoder(),
+            windowingStrategy.getWindowFn().windowCoder());
+
+    WindowedValue.
+            FullWindowedValueCoder<SingletonKeyedWorkItem<K, V>> windowedWorkItemCoder =
+            WindowedValue.getFullCoder(
+                    workItemCoder,
+                    windowingStrategy.getWindowFn().windowCoder());
+
+    CoderTypeInformation<WindowedValue<SingletonKeyedWorkItem<K, V>>> workItemTypeInfo =
+            new CoderTypeInformation<>(windowedWorkItemCoder);
+
+    DataStream<WindowedValue<SingletonKeyedWorkItem<K, V>>> workItemStream =
+            inputDataStream
+                    .flatMap(new FlinkStreamingTransformTranslators.ToKeyedWorkItem<>())
+                    .returns(workItemTypeInfo)
+                    .name("ToKeyedWorkItem");
+
+    KeyedStream<WindowedValue<SingletonKeyedWorkItem<K, V>>, ByteBuffer>
+            keyedWorkItemStream =
+            workItemStream.keyBy(new WorkItemKeySelector<>(inputCoder.getKeyCoder()));
+
+    SystemReduceFn<K, V, Iterable<V>, Iterable<V>, BoundedWindow> reduceFn =
+            SystemReduceFn.buffering(inputCoder.getValueCoder());
+
+    //Coder<WindowedValue<KV<K, Iterable<V>>>> outputCoder =
+    //        context.getCoder(context.getOutput(transform));
+    Coder<KV<K, Iterable<V>>> outputCoder;
+    try {
+      outputCoder = (Coder<KV<K, Iterable<V>>>) CoderTranslation.fromProto(
+              outputCoderProto,
+              rehydratedComponents);
+    } catch (IOException e) {
+      throw new IllegalStateException(
+              String.format("Unable to hydrate GroupByKey input coder %s.",
+                      inputCoderProto),
+              e);
+    }
+
+    //TypeInformation<WindowedValue<KV<K, Iterable<V>>>> outputTypeInfo =
+    //        context.getTypeInfo(context.getOutput(transform));
+    // TODO: use output windowing strategy?
+    WindowedValue.FullWindowedValueCoder<?> windowedValueOutputCoder =
+            WindowedValue.getFullCoder(
+                    outputCoder,
+                    windowingStrategy.getWindowFn().windowCoder());
+    TypeInformation<WindowedValue<KV<K, Iterable<V>>>> outputTypeInfo =
+            new CoderTypeInformation(windowedValueOutputCoder);
+
+    TupleTag<KV<K, Iterable<V>>> mainTag = new TupleTag<>("main output");
+
+    WindowDoFnOperator<K, V, Iterable<V>> doFnOperator =
+            new WindowDoFnOperator<K, V, Iterable<V>>(
```

Comment:
I think it's worth adding A TODO here to mention that we do not plan to ultimately wrap this step in a DoFn operator.

Reply:
done